### ISSUE TITLE
[Feature Request] Dex support in Jupyter notebooks.

### ISSUE THREAD

**Issue by dynamicwebpaige** [google-research/dex-lang#99]
![1_BIDimBvS_g8QH5fuHS_QRQ](https://user-images.githubusercontent.com/3712347/84825620-8c4dea00-afd6-11ea-8e86-383013d3dc53.gif)

**Way-far-down-the-road feature request:** 
But [Jupyter just recently added supported for SQLite](https://blog.jupyter.org/a-jupyter-kernel-for-sqlite-9549c5dcf551), and has support for specialized language kernels, such as [Haskell](https://mmhaskell.com/blog/2019/3/4/shareable-haskell-with-jupyter). Would it be possible to add support for Dex?

**Comment by oxinabox** [comment]
While Dex in Jupyter would indeed be cool,
if you haven't already checked it out Dex's own inbuild "notebook" interface I recommend it.
It is really cool and innovative.
The thing you get from running `dex web examples/tutorial.dx`.
It grew on me very quickly after worked out how it works.

Its very different to jupyter as you do the editting in a text-editor and the webpage just evaluates and renders it.
and the fact the Dex is a suitably pure functional language means it can be done stateless without being super slow.
Statefulness harming reproducability being the most common complaint about Jupyter in my experience ("Got to run these cells in this order").

Anyway, not a substitute for Jupyter, at least in terms of getting something people are already familar with out there,
but worth checking out.

**Comment by apaszke** [comment]
Thanks for the feature request! I agree with @oxinabox that our plan with Dex is actually to make a nicer alternative to Jupyter notebooks, based on the fact that we can design Dex to allow us to e.g. cache execution, show results incrementally and avoid the dependence on the order in which the cells were evaluated. However, Jupyter is so common that even if we're hoping that we can come up with something better, it might still make sense to integrate with that (Dex on Colab?).

**Comment by 8bitmp3** [comment]
> However, Jupyter is so common that even if we're hoping that we can come up with something better, it might still make sense to integrate with that (Dex on Colab?).

+1 👍

Just out of curiosity, have you discussed this with [at] colaboratory-team? Or would it make sense to have a separate thing like `swift-jupyter` that works as a sort of add-on (both in Jupyter and Colab)? 

_(From personal experience and after reading N number of StackOverflow comments/GH issues/How to Get Started blog posts  covering various ML projects, "how do I get started quickly" is still a major thing in 2020. So, in case there's a (not an immediate) priority at the Dex team to get outside user feedback, making the installation process less "complex" would be kind of awesome. Right now, we have to set up `stack`, then `llvm-9-dev` (or building from source) > `make`. If someone is not as curious, they may move on to another shiny thing :-). So, having a Jupyter integration in the short- to medium-term would probably reduce this initial friction, so new users/researchers can get straight to How To's, Tutorials and whatnot. Alternatively, maybe something as simple as `apt-get`, `brew` or `pip3` install?)_

**Comment by apaszke** [comment]
@8bitmp3 I agree 100% with your point about making the installation simple. However, Dex is at a point where I would certainly not consider it stable software, as even the surface level syntax of the language changes from time to time (there was actually quite a major shift that we've merged just last week). Once we're at a point, where we'll think that it makes sense to actually get more people onboard, we will definitely invest in making Dex installation a breeze! But of course if you or anyone else is willing to try putting together a Jupyter extension, then we welcome contributions and will be happy to help you get that done!

**Comment by 8bitmp3** [comment]
Hey @apaszke Sorry, a bit off-topic here but it's related to installation and I didn't want to spam your team with another installation "issue". In the macOS setup, installing `LLVM 9` is a bit tricky and your team shared this discussion in the instructions: https://github.com/google-research/dex-lang/issues/2#issuecomment-649896955. Do you think it would make sense to also include https://github.com/google-research/dex-lang/issues/2#issuecomment-557793009 (How to build and run dex on Mac - updated 20 days ago)? Thanks!

> Once we're at a point, where we'll think that it makes sense to actually get more people onboard, we will definitely invest in making Dex installation a breeze!

Looking forward to that 🚀

**Comment by apaszke** [comment]
Now that we have a reasonable Python integration, this would actually be a very cool thing to have!

**Comment by 8bitmp3** [comment]
![](https://cultofthepartyparrot.com/parrots/hd/reverseportalblueparrot.gif)


![](https://cultofthepartyparrot.com/parrots/hd/portalorangeparrot.gif)

**Comment by pharringtonp19** [comment]
Out of curiosity, what is the latest with this thinking/plan?

**Comment by apaszke** [comment]
No updates or plans at the moment, perhaps except for a willingness to design and review external contributions 🙃
